### PR TITLE
Support RelationList field with StaticCatalogVocabulary and SelectWidget.

### PR DESCRIPTION
**Pull request by ksuess**
**Please test with https://github.com/plone/plone.restapi/pull/1605**

---
See example.contenttype https://github.com/collective/example.contenttype for example content type with relation fields for testing purpose.

image

---

See also https://github.com/plone/volto/issues/3355

### CODE REVIEW

**Comment by ksuess** [comment]
How To:
https://deploy-preview-4614--volto.netlify.app/recipes/widget.html#relation-fields

**Comment by ksuess** [comment]
Please test with https://github.com/plone/plone.restapi/pull/1605

File: docs/source/recipes/widget.md [hunk 1]
```diff
@@ -146,14 +146,102 @@ const applyConfig = (config) => {
 Based on this setup, Volto will render this field with the `TokenWidget`.
 
 
-```{seealso}
-See [storybook](https://6.dev-docs.plone.org/storybook) with available widgets.
+## Relation fields
```
Comment (stevepiercy) [inline docs/source/recipes/widget.md:149]: Section headings without narrative text should be avoided. They should either:

- have introductory general text describing their sub-sections, or
- be removed and their sub-sections demoted.

File: docs/source/recipes/widget.md [hunk 2]
```diff
@@ -146,14 +146,102 @@ const applyConfig = (config) => {
...
+
+### Single relation field
+
+Relation field with named `StaticCatalogVocabulary`and `Select` widget:
```
Comment (stevepiercy) [inline docs/source/recipes/widget.md:153]:
```suggestion
The following example is a single relation choice field with the named vocabulary of `StaticCatalogVocabulary` as a `Select` widget:
```

File: docs/source/recipes/widget.md [hunk 3]
```diff
@@ -146,14 +146,102 @@ const applyConfig = (config) => {
...
+    "relationchoice_field_named_staticcatalogvocabulary",
+    frontendOptions={
+        "widget": "select",
+    },
+)
+```
+
+
+It is recommended to define the vocabulary as a named `StaticCatalogVocabulary` with the field/relation name as its name.  
+This allowes the relations control panel to respect the defined restrictions to potential relation targets.
```
Comment (stevepiercy) [inline docs/source/recipes/widget.md:172]:
```suggestion
This allows the {guilabel}`relations` control panel to respect the defined restrictions to potential relation targets.
```

File: docs/source/recipes/widget.md [hunk 4]
```diff
@@ -146,14 +146,102 @@ const applyConfig = (config) => {
...
+)
+directives.widget(
+    "relationchoice_field_named_staticcatalogvocabulary",
+    frontendOptions={
+        "widget": "select",
+    },
+)
+```
+
+
```
Comment (stevepiercy) [inline docs/source/recipes/widget.md:170]:
```suggestion
```

File: docs/source/recipes/widget.md [hunk 5]
```diff
@@ -146,14 +146,102 @@ const applyConfig = (config) => {
...
 
+{file}`configure.zcml`
+```xml
+<utility
+  name="relationchoice_field_named_staticcatalogvocabulary"
+  component="example.contenttype.vocabularies.ExamplesVocabularyFactory"
+  />
+```
+
+
```
Comment (stevepiercy) [inline docs/source/recipes/widget.md:200]:
```suggestion
```

File: docs/source/recipes/widget.md [hunk 6]
```diff
@@ -146,14 +146,102 @@ const applyConfig = (config) => {
...
+The directive is by now default and can be ommitted.
```
Comment (stevepiercy) [inline docs/source/recipes/widget.md:201]:
```suggestion
By following this guidance, the directive is set by default and can be omitted as shown below.
```

File: docs/source/recipes/widget.md [hunk 7]
```diff
@@ -146,14 +146,102 @@ const applyConfig = (config) => {
...
+    description="field/relation: relationchoice_field_named_staticcatalogvocabulary",
+    vocabulary="relationchoice_field_named_staticcatalogvocabulary",
+    required=False,
+)
+```
+
+
+### Multi relation field
+
+Multi relation field with named `StaticCatalogVocabulary`and `Select` widget:
```
Comment (stevepiercy) [inline docs/source/recipes/widget.md:215]:
```suggestion
Similar to {ref}`single-relation-field-label`, the following example is a multi relation choice field with the named vocabulary of `StaticCatalogVocabulary` as a `Select` widget:
```

File: docs/source/recipes/widget.md [hunk 8]
```diff
@@ -146,14 +146,102 @@ const applyConfig = (config) => {
 Based on this setup, Volto will render this field with the `TokenWidget`.
 
 
-```{seealso}
-See [storybook](https://6.dev-docs.plone.org/storybook) with available widgets.
+## Relation fields
+
```
Comment (stevepiercy) [inline docs/source/recipes/widget.md:150]:
```suggestion

(single-relation-field-label)=

```

File: docs/source/recipes/widget.md [hunk 9]
```diff
@@ -146,14 +146,102 @@ const applyConfig = (config) => {
...
+```python
+relationchoice_field_named_staticcatalogvocabulary = RelationChoice(
+    title="RelationChoice – named StaticCatalogVocabulary – Select widget",
+    description="field/relation: relationchoice_field_named_staticcatalogvocabulary",
+    vocabulary="relationchoice_field_named_staticcatalogvocabulary",
+    required=False,
+)
+```
+
+
```
Comment (stevepiercy) [inline docs/source/recipes/widget.md:212]:
```suggestion

(multi-relation-field-label)=

```

**Review by stevepiercy — CHANGES_REQUESTED** [review]
A few enhancements, grammar and spelling fixes, and suggestions.

I am not sure if I got the introductory descriptions of each relation field exactly correct. Feel free to revise them to be more accurate.

File: docs/source/recipes/widget.md [hunk 10]
```diff
@@ -146,14 +146,112 @@ const applyConfig = (config) => {
 Based on this setup, Volto will render this field with the `TokenWidget`.
 
 
-```{seealso}
-See [storybook](https://6.dev-docs.plone.org/storybook) with available widgets.
+(widget-relation-field-label)=
+
+## Relation fields
+
+A relation field is either a single relation field to hold at most one content object, `RelationChoice`, or a multi relation field `RelationList`, that can hold more than one content object.
```
Comment (stevepiercy) [inline docs/source/recipes/widget.md:153]:
```suggestion
A relation field is either a single relation field to hold at most one content object, `RelationChoice`, or a multi relation field, `RelationList`, that can hold more than one content object.
```

File: docs/source/recipes/widget.md [hunk 11]
```diff
@@ -146,14 +146,112 @@ const applyConfig = (config) => {
...
+
+Relation fields can be edited and rendered with the `Select` widget.
+The restriction on content types, workflow states, e.a. can be done with a `StaticCatalogVocabulary`.
```
Comment (stevepiercy) [inline docs/source/recipes/widget.md:156]:
```suggestion
The restriction on content types, workflow states, and so on can be done with a `StaticCatalogVocabulary`.
```

File: docs/source/recipes/widget.md [hunk 12]
```diff
@@ -146,14 +146,112 @@ const applyConfig = (config) => {
...
+
+There are other vocabulary types and other widgets like the `ObjectBrowser` widget.
```
Comment (stevepiercy) [inline docs/source/recipes/widget.md:158]:
```suggestion
There are other vocabulary types and other widgets, including the `ObjectBrowser` widget.
```

File: docs/source/recipes/widget.md [hunk 13]
```diff
@@ -146,14 +146,112 @@ const applyConfig = (config) => {
...
+{file}`configure.zcml`
+```xml
+<utility
+  name="relationchoice_field_named_staticcatalogvocabulary"
+  component="example.contenttype.vocabularies.ExamplesVocabularyFactory"
+  />
+```
+
+The `Select` widget is currently the default for `RelationChoice` fields with vocabulary.
+Therefore the directive can be ommitted.
```
Comment (stevepiercy) [inline docs/source/recipes/widget.md:210]:
```suggestion
Therefore the directive can be omitted.
```

Comment (stevepiercy) [inline docs/source/recipes/widget.md:210]:
```suggestion
Therefore the directive can be omitted.
```

**Review by stevepiercy — CHANGES_REQUESTED** [review]
Nice revision. This is much clearer. Minor spelling and grammar fixes, and the docs are good to go. Thank you!

**Review by stevepiercy — APPROVED** [review]
Docs approved. Code should be reviewed by a core maintainer.

**Review by tiberiuichim — APPROVED** [review]
Awesome work @ksuess !

**Review by sneridagh — APPROVED** [review]
LGTM! Awesome work @ksuess !!! Very much needed!

I'll merge as soon we have a release in p.restapi. I'll push for it.

**Pull request merged**